pkg/readiness/cert_tracker.go

```diff
@@ -0,0 +1,112 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package readiness
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"io/ioutil"
+	"net/http"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var ctLog = logf.Log.WithName("cert-tracker")
+
+// CertTracker tracks readiness for certs of the webhook.
+type CertTracker struct {
+	certDir string
+	dnsName string
+}
+
+// NewCertTracker creates a new CertTracker
+func NewCertTracker(certDir string, dnsName string) *CertTracker {
+	return &CertTracker{
+		certDir: certDir,
+		dnsName: dnsName,
+	}
+}
+
+// CheckCert implements healthz.Checker to report readiness based on cert validity
+// the readiness probe returns nil if valid, otherwise returns an error.
+func (c *CertTracker) CheckCert(req *http.Request) error {
+	ctLog.V(1).Info("readiness checker CheckCert started")
+
+	// Load files
+	caCrt, err := ioutil.ReadFile(filepath.Join(c.certDir, "ca.crt"))
```

Comment:
This check performs file I/O and a bunch of crypto - do we have any concerns around the probe being a DoS vector for Gatekeeper and by proxy the API Server?
* Should we cache certs and detect changes to the source files before rereading?
* Should we cache parsed certs and only re-run verification?
* Should we rate-limit the check endpoint to mitigate?
* ¯\_(ツ)_/¯

Reply:
If we do cache, how often do we refresh the cache such that the maximum duration is an acceptable period during which the pod is still ready while the cert is no longer valid.

Reply:
A few thoughts:

One option would be to monitor the filesystem for changes: https://github.com/fsnotify/fsnotify
Another would be to let Kubernetes tell use when the secrets change.
A third would be to implement an arbitrary timeout or a timeout based on certificate expiration.
A fourth might be to rate-limit the file refresh part and to emit the last response if the limit is reached.

Do you think this is a valid concern and worth additional complexity?

Reply:
It'd be cool if we could listen in on certWatcher

https://github.com/kubernetes-sigs/controller-runtime/blob/c000ea850121b53dd769f624f3bfa74a4fcf100f/pkg/webhook/server.go#L197

though I don't know if that's possible.

re: crypto complexity....

I'm not sure how much we have to worry about that, aren't we doing crypto operations for every HTTP request anyway? I don't know if we're leveraging things like session keys to lower the overhead. Disk usage could be a concern.

If crypto complexity is a concern, maybe we should only do the readiness check until we are up? Really the test of whether the cert is valid from the API server's point of view requires making sure we are consistent with the ValidatingWebhookConfiguration, so we are approximating anyway.

Finally, what exactly are we trying to solve for here? The server isn't running until a valid cert is loaded anyway. Are we looking for cases where the TLS cert becomes invalid? If we are okay with locking the tracker to be hard-available as the above paragraph suggests, then the tracker itself might be redundant.

Reply:
> what exactly are we trying to solve for here?

This is to update the pod as not ready when the cert becomes invalid post the initial start of the deployment. Imagine the operator generated their own cert for the webhook and after sometime, it expires. The pods are still in ready state while all subsequent webhook communication with the api server are now all failing. This new readiness checker will mark these pods as not ready as soon as the certs are no longer valid, thus signal the unhealthy state of GK to anything that is observing its health.

Reply:
Ack, we should probably validate the cert against the validatingwebhookconfiguration object if we want to be faithful to the question "is this cert valid?".

Because we already have controllers that try to enforce a valid cert condition, It's possible we may actually be hurting uptime by creating redundant systems. I would expect the following sequence of events using this code:

cert invalid -> rejected requests due to cert invalidity for time Tx -> webhook marked unhealthy -> rejected requests due to webhook unavailability for time Ty -> cert fixed -> lag time for time Tz during which the cert is valid but the webhook is still marked unhealthy -> webhook marked healthy

In the case where we don't have this check, the rejected requests time is Tx + Ty, and there is no Tz. I don't know the relative size of any of the times (that'd need to be discovered empirically), but that they exist is very likely.

We know two things:

`(Tx + Ty + Tz) >= (Tx + Ty)`

This means that a webhook will have greater overall down time by adding this check.

`(Tx + Ty) >= Tx`

This means that we run the risk of having requests rejected because of a bad cert as opposed to (potentially) being routed to a more well-behaved host.

There is a question of whether downing the webhooks actually yield a practical benefit in terms of uptime. I'm not sure it does. Because all webhooks use the same cert and listen to the same secret, they will likely all succeed or fail together. The uptime benefit of marking a Pod as "unhealthy" is the prospect that there are other healthy pods that would serve the request.

I think we are faced with two questions:

1. Do we get any benefit from failing due to no healthy webhooks vs. failing due to bad certs given the SPOF certs are to the entire system?
2.  If we get any benefit, what are the relative values of Tx, Ty, Tz? Are the costs of potential extra downtime and code complexity worth the benefits of uptime.

IMO (2) can't be answered without writing this code and experimenting with it. (1) sounds like something we can address without merging. I'm curious to hear people's thoughts on it.

If the answer to (1) is no benefit, we would need to focus on eliminating the SPOF. It's not clear to me how we would do this unless webhook configs allowed for more than one root cert so that Pods could gradually rotate to a new cert. At that point code like this could be very useful, as drift in cert validity between pods becomes possible.